protoarray/src/main/java/tech/pegasys/artemis/protoarray/ProtoArrayForkChoiceStrategy.java

```diff
@@ -113,30 +119,30 @@ public void maybePrune(Bytes32 finalizedRoot) {
 
   // Internal
 
-  static ProtoArrayForkChoiceStrategy create(
-      UnsignedLong finalizedBlockSlot,
-      Bytes32 finalizedBlockStateRoot,
-      UnsignedLong justifiedEpoch,
-      UnsignedLong finalizedEpoch,
-      Bytes32 finalizedBlockRoot) {
-    ProtoArray protoArray =
-        new ProtoArray(
-            Constants.PROTOARRAY_FORKCHOICE_PRUNE_THRESHOLD,
-            justifiedEpoch,
-            finalizedEpoch,
-            new ArrayList<>(),
-            new HashMap<>());
+  private static void processBlocksInStoreAtStartup(ReadOnlyStore store, ProtoArray protoArray) {
+    List<BeaconBlock> blocks = new ArrayList<>();
+    for (Bytes32 blockRoots : store.getBlockRoots()) {
+      BeaconBlock block =
+          checkNotNull(
+              store.getBlock(blockRoots),
+              "ProtoArrayForkChoiceStrategy: Store does not contain the block of the block root it stores");
+      blocks.add(block);
+    }
 
-    protoArray.onBlock(
-        finalizedBlockSlot,
-        finalizedBlockRoot,
-        Optional.empty(),
-        finalizedBlockStateRoot,
-        justifiedEpoch,
-        finalizedEpoch);
+    blocks.sort(Comparator.comparing(BeaconBlock::getSlot));
 
-    return new ProtoArrayForkChoiceStrategy(
-        protoArray, new ElasticList<>(VoteTracker::Default), new ArrayList<>());
+    for (BeaconBlock block : blocks) {
+      Bytes32 blockRoot = block.hash_tree_root();
+      protoArray.onBlock(
+          block.getSlot(),
+          blockRoot,
+          store.getBlockRoots().contains(block.getParent_root())
+              ? Optional.of(block.getParent_root())
+              : Optional.empty(),
+          block.getState_root(),
+          store.getBlockState(block.hash_tree_root()).getCurrent_justified_checkpoint().getEpoch(),
+          store.getBlockState(block.hash_tree_root()).getFinalized_checkpoint().getEpoch());
+    }
   }
```

Comment:
You can simplify this with streams to:

```
  private static void processBlocksInStoreAtStartup(ReadOnlyStore store, ProtoArray protoArray) {
    store.getBlockRoots().stream()
        .map(store::getBlock)
        .sorted(Comparator.comparing(BeaconBlock::getSlot))
        .forEach(block -> processBlockAtStartup(store, protoArray, block));
  }

  private static void processBlockAtStartup(
      final ReadOnlyStore store, final ProtoArray protoArray, final BeaconBlock block) {
    Bytes32 blockRoot = block.hash_tree_root();
    protoArray.onBlock(
        block.getSlot(),
        blockRoot,
        store.getBlockRoots().contains(block.getParent_root())
            ? Optional.of(block.getParent_root())
            : Optional.empty(),
        block.getState_root(),
        store.getBlockState(block.hash_tree_root()).getCurrent_justified_checkpoint().getEpoch(),
        store.getBlockState(block.hash_tree_root()).getFinalized_checkpoint().getEpoch());
  }
```

Reply:
You could put the `checkNotNull` back in but I'm not sure it's worth it given that `Store` guarantees that it has a block for every block root (and it literally gets the block roots from the keyset of its map of blocks).

Reply:
This is neat. Thank you.